providers/providers_manager.go

```diff
@@ -250,6 +250,42 @@ func (pm *ProviderManager) getProvidersForKey(k []byte) ([]peer.ID, error) {
 	return pset.providers, nil
 }
 
+// GetAllProviders returns map of all providers where key is the provide key base32
+// encoded and the value is a slice of peerIDs that we know to be providing the key
+func (pm *ProviderManager) GetAllProviders() (map[string][]peer.ID, error) {
```

Comment:
What if this was `GetAllProviderKeys` that returned a `chan string`? You could keep track of seen keys and the caller code could receive a provider key and call `GetProviders` for each?

Reply:
That would yield a database read for each provider for a GetAllProviders vs. one single read for all of them (unless some levelDB implementation detail that I'm missing)

Reply:
This should stream results back. Either key/provider tuples (`struct {key string, provider peer.ID}`) or a struct with all providers for a given key `struct { key string, providers []peer.ID }`.

Reply:
> That would yield a database read for each provider for a GetAllProviders vs. one single read for all of them (unless some levelDB implementation detail that I'm missing)

True, that was short signted. I was trying to think of a solution to buffering the entire datastore in memory. Unless pre-sorted, `struct {key string, provider peer.ID}` would just mean the caller will end up buffering the whole channel in order to group all the providers per key.

`struct { key string, providers []peer.ID }` is really what we want. Is possible without buffering everything in memory for the sort?

Reply:
You can ask for the datastore query to be sorted. On most datastores, that will be free.